<?xml version="1.0" encoding="UTF-8"?>

<mule xmlns:db="http://www.mulesoft.org/schema/mule/db" xmlns:validation="http://www.mulesoft.org/schema/mule/validation"
	xmlns:ee="http://www.mulesoft.org/schema/mule/ee/core"
	xmlns:http="http://www.mulesoft.org/schema/mule/http" xmlns="http://www.mulesoft.org/schema/mule/core" xmlns:doc="http://www.mulesoft.org/schema/mule/documentation" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mulesoft.org/schema/mule/core http://www.mulesoft.org/schema/mule/core/current/mule.xsd
http://www.mulesoft.org/schema/mule/http http://www.mulesoft.org/schema/mule/http/current/mule-http.xsd
http://www.mulesoft.org/schema/mule/ee/core http://www.mulesoft.org/schema/mule/ee/core/current/mule-ee.xsd
http://www.mulesoft.org/schema/mule/validation http://www.mulesoft.org/schema/mule/validation/current/mule-validation.xsd
http://www.mulesoft.org/schema/mule/db http://www.mulesoft.org/schema/mule/db/current/mule-db.xsd">
	<sub-flow name="client-cca-call-services-get-order-remision" doc:id="ec19fe54-08e9-4fea-8454-6cc2ec8082e3" >
		<logger level="INFO" doc:name="Logger" doc:id="1d2450f6-d3b6-48f8-b9cb-31517a47a88e" message="#[payload]"/>
			<try doc:name="Try" doc:id="cca68d64-0533-4543-b725-c52921100323" >
			<until-successful maxRetries="5" doc:name="Until Successful" doc:id="1fcf1678-803b-4ba5-910e-0f0a966a806b" millisBetweenRetries="30000">
				<http:request method="GET" doc:name="Request" doc:id="fc6f1876-9dc1-4ecf-bbc8-0e47b7803e25" config-ref="Http_request_CCA" path="${https.request.cca.api.get.order.remision}">
			<http:headers><![CDATA[#[output application/java
---
{
	"Host" : p('https.request.cca.api.host')
}]]]></http:headers>
			<http:uri-params><![CDATA[#[output application/java
---
{
	"orden" : vars.orden
}]]]></http:uri-params>
			<http:query-params><![CDATA[#[output application/java
---
{
	"clasPtal" : vars.clasPtal,
	"contrato" : vars.contrato,
	"password" : vars.password
}]]]></http:query-params>
		</http:request>
			</until-successful>
			<error-handler >
				<on-error-continue enableNotifications="true" logException="true" doc:name="On Error Continue" doc:id="8254d1b5-f724-49df-b217-76e30d419efe" type="CCA:BAD_REQUEST, HTTP:BAD_REQUEST, HTTP:INTERNAL_SERVER_ERROR">
					<ee:transform doc:name="Transform Message" doc:id="ac92ff4a-6f22-417e-bcf7-8c0e96ec91c5">
						<ee:message>
						</ee:message>
						<ee:variables>
							<ee:set-variable variableName="errorCode"><![CDATA[%dw 2.0
output application/json
---
error.errorMessage.payload.code]]></ee:set-variable>
							<ee:set-variable variableName="errorMessage"><![CDATA[%dw 2.0
output application/json
---
write(error.errorMessage.payload,"application/json")]]></ee:set-variable>
						</ee:variables>
					</ee:transform>
					<validation:is-false doc:name="Is false" doc:id="0517bea4-463d-4e40-babf-8a2a3fd6bfac" expression='#[vars.errorCode == "409"]' message="#[vars.errorMessage]">
						<error-mapping sourceType="VALIDATION:INVALID_BOOLEAN" targetType="CCA:BAD_REQUEST" />
					</validation:is-false>
				</on-error-continue>
				<on-error-continue enableNotifications="true" logException="true" doc:name="On Error Continue" doc:id="0b0f0159-ebb9-496e-98bb-9ffe61c8012d" type="HTTP:CLIENT_SECURITY, HTTP:CONNECTIVITY, HTTP:FORBIDDEN, HTTP:RETRY_EXHAUSTED, HTTP:SERVICE_UNAVAILABLE, HTTP:TIMEOUT, HTTP:UNAUTHORIZED">
					<ee:transform doc:name="Transform Message" doc:id="1f2ba610-cb64-4963-8d5b-95ab03c74f97">
						<ee:message />
						<ee:variables>
							<ee:set-variable variableName="causeby"><![CDATA[%dw 2.0
output application/json
---
error.description]]></ee:set-variable>
							<ee:set-variable variableName="main_error"><![CDATA[%dw 2.0
output application/json
---
error.errorType]]></ee:set-variable>
						</ee:variables>
					</ee:transform>
					<ee:transform doc:name="set payload error" doc:id="69f0a3ec-c899-4667-9770-18add8eaedea">
						<ee:message>
							<ee:set-payload><![CDATA[%dw 2.0
output application/json
---
{
  traxion_response: {
    completed_succesfully: "false",
    error: {
      error_type: "CCA:" ++ (vars.main_error.identifier default "ERROR"),
      payload: {
      	event: correlationId
      },
      user_error_description: "error en comunicacion con CCA",
      system_error_description: vars.causeby
    }
  }
}]]></ee:set-payload>
						</ee:message>
						<ee:variables>
							<ee:set-variable variableName="slackChannel"><![CDATA[%dw 2.0
output application/json
---
p('slack.connection.channel.connectivity.error')]]></ee:set-variable>
							<ee:set-variable variableName="messageProvider"><![CDATA[%dw 2.0
output application/json
---
"ERROR PROVENIENTE DE WMS BLUEYONDER"]]></ee:set-variable>
						</ee:variables>
					</ee:transform>
					<flow-ref doc:name="orchestrator-send-message-to-slack-main" doc:id="e98f0606-0bb2-419c-920e-e0b539bbaf1d" name="orchestrator-send-message-to-slack-main" />
					<raise-error doc:name="Raise error" doc:id="33195c12-c054-439a-b25c-2a362e5a5f62" type="CCA:ERROR" description="#[vars.causeby]" />
				</on-error-continue>
			</error-handler>
		</try>
		<logger level="INFO" doc:name="Logger" doc:id="82a98730-5fe0-4c9a-920f-c7bf519d506c" message="#[payload]" />
		<ee:transform doc:name="prepare response" doc:id="3a1027f9-159a-464c-b3a8-39c22ef6331c">
			<ee:message>
				<ee:set-payload><![CDATA[%dw 2.0
output application/json
fun validateStatus(estatusOrden) =
estatusOrden match {
  case "0" -> { "estatusDescription": "GENERADA/A. PARCIAL"}
  case "1" -> { "estatusDescription": "INCUMPLIDA"}
  case "2" -> { "estatusDescription": "ATENDIDA ( TOTAL )"}
  case "4" -> { "estatusDescription": "CANCELADA"}
  else -> { "estatusDescription": "NO DISPONIBLE"}
}
---
payload ++ validateStatus(payload.estatusOrden)]]></ee:set-payload>
			</ee:message>
		</ee:transform>
	</sub-flow>
	<sub-flow name="client-cca-call-services-get-order-remision-bienestar" doc:id="b79be803-f372-4b15-a094-3469394b2a43" >
		<logger level="INFO" doc:name="Logger" doc:id="32851433-eaa8-4952-9f4c-b217781147fe" message="#[payload]"/>
		<try doc:name="Try" doc:id="00c26735-4575-4609-8325-c3c976922b08" >
			<until-successful maxRetries="5" doc:name="Until Successful" doc:id="ef5ba55f-711d-4d5b-a7c6-3f2a6bb0c0b7" millisBetweenRetries="30000">
				<http:request method="GET" doc:name="Request" doc:id="3f71df19-978a-47e3-8307-a73548a04448" config-ref="Http_request_CCA_Bienestar" path="${https.request.cca.bienestar.api.get.order.remision}">
			<http:headers><![CDATA[#[output application/java
---
{
	"Host" : p('https.request.cca.bienestar.api.host')
}]]]></http:headers>
			<http:uri-params><![CDATA[#[output application/java
---
{
	"orden" : vars.orden
}]]]></http:uri-params>
			<http:query-params><![CDATA[#[output application/java
---
{
	"clasPtal" : vars.clasPtal,
	"contrato" : vars.contrato,
	"password" : vars.password
}]]]></http:query-params>
		</http:request>
			</until-successful>
			<error-handler >
				<on-error-continue enableNotifications="true" logException="true" doc:name="On Error Continue" doc:id="666ea41c-bcf8-4763-b8de-3d1408045ae6" type="CCA:BAD_REQUEST, HTTP:BAD_REQUEST, HTTP:INTERNAL_SERVER_ERROR">
					<ee:transform doc:name="Transform Message" doc:id="025bcbf2-de9a-4b73-a49f-4c2d5c571c0b" >
						<ee:message />
						<ee:variables >
							<ee:set-variable variableName="errorCode" ><![CDATA[%dw 2.0
output application/json
---
error.errorMessage.payload.code]]></ee:set-variable>
							<ee:set-variable variableName="errorMessage" ><![CDATA[%dw 2.0
output application/json
---
write(error.errorMessage.payload,"application/json")]]></ee:set-variable>
						</ee:variables>
					</ee:transform>
					<validation:is-false doc:name="Is false" doc:id="bce938ab-9bfa-4ddf-b187-748195bf15bb" expression='#[vars.errorCode == "409"]' message="#[vars.errorMessage]" >
						<error-mapping sourceType="VALIDATION:INVALID_BOOLEAN" targetType="CCA:BAD_REQUEST" />
					</validation:is-false>
				</on-error-continue>
				<on-error-continue enableNotifications="true" logException="true" doc:name="On Error Continue" doc:id="a435a066-08b6-44ea-a5b1-761b7efbe58a" type="HTTP:CLIENT_SECURITY, HTTP:CONNECTIVITY, HTTP:FORBIDDEN, HTTP:RETRY_EXHAUSTED, HTTP:SERVICE_UNAVAILABLE, HTTP:TIMEOUT, HTTP:UNAUTHORIZED">
					<ee:transform doc:name="Transform Message" doc:id="63b95141-8484-4004-be5d-88c8b4ad6139" >
						<ee:message />
						<ee:variables >
							<ee:set-variable variableName="causeby" ><![CDATA[%dw 2.0
output application/json
---
error.description]]></ee:set-variable>
							<ee:set-variable variableName="main_error" ><![CDATA[%dw 2.0
output application/json
---
error.errorType]]></ee:set-variable>
						</ee:variables>
					</ee:transform>
					<ee:transform doc:name="set payload error" doc:id="0093bff5-d249-44b8-98bd-1b43eb79903b" >
						<ee:message >
							<ee:set-payload ><![CDATA[%dw 2.0
output application/json
---
{
  traxion_response: {
    completed_succesfully: "false",
    error: {
      error_type: "CCA:" ++ (vars.main_error.identifier default "ERROR"),
      payload: {
      	event: correlationId
      },
      user_error_description: "error en comunicacion con CCA",
      system_error_description: vars.causeby
    }
  }
}]]></ee:set-payload>
						</ee:message>
						<ee:variables >
							<ee:set-variable variableName="slackChannel" ><![CDATA[%dw 2.0
output application/json
---
p('slack.connection.channel.connectivity.error')]]></ee:set-variable>
							<ee:set-variable variableName="messageProvider" ><![CDATA[%dw 2.0
output application/json
---
"ERROR PROVENIENTE DE WMS BLUEYONDER"]]></ee:set-variable>
						</ee:variables>
					</ee:transform>
					<flow-ref doc:name="orchestrator-send-message-to-slack-main" doc:id="bc55115a-920b-4f16-bd16-b55c088ab098" name="orchestrator-send-message-to-slack-main" />
					<raise-error doc:name="Raise error" doc:id="b553101a-a0a1-4a63-83e0-eb6d969bdced" type="CCA:ERROR" description="#[vars.causeby]" />
				</on-error-continue>
			</error-handler>
		</try>
		<logger level="INFO" doc:name="Logger" doc:id="145d0464-1543-41b3-a9a4-cf1a841dfc0f" message="#[payload]"/>
	</sub-flow>
	<sub-flow name="client-cca-call-services-get-order-confirmation" doc:id="93809432-409f-4b12-b995-26a263a938a4" >
		<logger level="INFO" doc:name="Logger" doc:id="fced381f-8cb9-47ba-9ad5-337179b99590" message="#[payload]"/>
		<try doc:name="Try" doc:id="5f2b6c32-bd43-4e3b-b94c-0a13db5a5c60" >
			<until-successful maxRetries="5" doc:name="Until Successful" doc:id="5955c611-8fa2-489b-8fe0-07132c6a5c99" millisBetweenRetries="30000">
			<http:request method="POST" doc:name="Request" doc:id="8cc5325d-88e3-4bbf-870c-1de232e2d6ac" config-ref="Http_request_CCA" path="${https.request.cca.api.post.create.remision}">
			<http:headers><![CDATA[#[output application/java
---
{
	"Host" : p('https.request.cca.api.host')
}]]]></http:headers>
		</http:request>
		</until-successful>
			<error-handler >
				<on-error-continue enableNotifications="true" logException="true" doc:name="On Error Continue" doc:id="617bbf96-c6bc-4c85-8c2f-deff27dd2b50" type="HTTP:CLIENT_SECURITY, HTTP:CONNECTIVITY, HTTP:FORBIDDEN, HTTP:RETRY_EXHAUSTED, HTTP:SERVICE_UNAVAILABLE, HTTP:TIMEOUT, HTTP:UNAUTHORIZED">
					<ee:transform doc:name="Transform Message" doc:id="1af5f3f1-27ff-46dd-9d80-d932efefc308" >
						<ee:message />
						<ee:variables >
							<ee:set-variable variableName="causeby" ><![CDATA[%dw 2.0
output application/json
---
error.description]]></ee:set-variable>
							<ee:set-variable variableName="main_error" ><![CDATA[%dw 2.0
output application/json
---
error.errorType]]></ee:set-variable>
						</ee:variables>
					</ee:transform>
					<ee:transform doc:name="set payload error" doc:id="72cc46d0-3c4f-472a-9ff3-2216ec7af509" >
						<ee:message >
							<ee:set-payload ><![CDATA[%dw 2.0
output application/json
---
{
  traxion_response: {
    completed_succesfully: "false",
    error: {
      error_type: "CCA:" ++ (vars.main_error.identifier default "ERROR"),
      payload: {
      	event: correlationId
      },
      user_error_description: "error en comunicacion con CCA",
      system_error_description: vars.causeby
    }
  }
}]]></ee:set-payload>
						</ee:message>
						<ee:variables >
							<ee:set-variable variableName="slackChannel" ><![CDATA[%dw 2.0
output application/json
---
p('slack.connection.channel.connectivity.error')]]></ee:set-variable>
							<ee:set-variable variableName="messageProvider" ><![CDATA[%dw 2.0
output application/json
---
"ERROR PROVENIENTE DE WMS BLUEYONDER"]]></ee:set-variable>
						</ee:variables>
					</ee:transform>
					<flow-ref doc:name="orchestrator-send-message-to-slack-main" doc:id="4f4ac94d-8768-4487-942a-678c2f63c809" name="orchestrator-send-message-to-slack-main" />
					<raise-error doc:name="Raise error" doc:id="117c8058-f02d-4714-8037-c0423470511b" type="CCA:ERROR" description="#[vars.causeby]" />
				</on-error-continue>
			</error-handler>
		</try>
		<ee:transform doc:name="Catch http status 206" doc:id="6149e312-9df5-442a-8742-5c565e877b77">
				<ee:message>
				</ee:message>
				<ee:variables>
				<ee:set-variable variableName="codeErrorCca" ><![CDATA[attributes.statusCode]]></ee:set-variable>
				</ee:variables>
			</ee:transform>
		<validation:is-false doc:name="Is false" doc:id="0dc049da-ed2f-416e-82b2-b2b60e51edc8" expression="#[(!isEmpty(payload[0].erroresM)) or (!isEmpty(payload[0].erroresD))]" message='#[write(flatten(payload.erroresM ++ payload.erroresD),"application/json")]' >
			<error-mapping sourceType="VALIDATION:INVALID_BOOLEAN" targetType="CCA:BAD_REQUEST" />
		</validation:is-false>
		<logger level="INFO" doc:name="Logger" doc:id="142d806b-46a3-4c03-8ff6-e65eaf084b15" message="#[payload]"/>
	</sub-flow>
	<sub-flow name="client-cca-call-services-get-order-confirmation-bienestar" doc:id="b9fa3fd7-65a2-4a2a-b9d2-2eaad2452500" >
		<logger level="INFO" doc:name="Logger" doc:id="fdd2a906-56e1-4344-a9b7-74f65352d347" message="#[payload]"/>
		<try doc:name="Try" doc:id="343a9642-f0e1-4163-9c1f-7899ff868260">
				<until-successful maxRetries="5" doc:name="Until Successful" doc:id="1ea75314-87f9-4fb6-b08d-c7f85e1e3ac9" millisBetweenRetries="30000">
			<http:request method="POST" doc:name="Request" doc:id="7a241a90-c547-4b74-9134-d3bb2750901b" config-ref="Http_request_CCA_Bienestar" path="${https.request.cca.bienestar.api.post.create.remision}">
			<http:headers><![CDATA[#[output application/java
---
{
	"Host" : p('https.request.cca.bienestar.api.host')
}]]]></http:headers>
		</http:request>
		</until-successful>
			<error-handler>
					<on-error-continue enableNotifications="true" logException="true" doc:name="On Error Continue" doc:id="7bda275d-bc56-4a05-b7a0-e8e44e673abc" type="HTTP:CLIENT_SECURITY, HTTP:CONNECTIVITY, HTTP:FORBIDDEN, HTTP:RETRY_EXHAUSTED, HTTP:SERVICE_UNAVAILABLE, HTTP:TIMEOUT, HTTP:UNAUTHORIZED" >
					<ee:transform doc:name="Transform Message" doc:id="a5708d62-5051-462b-8aa7-d4dd8035a0c1" >
						<ee:message />
						<ee:variables >
							<ee:set-variable variableName="causeby" ><![CDATA[%dw 2.0
output application/json
---
error.description]]></ee:set-variable>
							<ee:set-variable variableName="main_error" ><![CDATA[%dw 2.0
output application/json
---
error.errorType]]></ee:set-variable>
						</ee:variables>
					</ee:transform>
					<ee:transform doc:name="set payload error" doc:id="ea5bb81e-0dfb-4b2f-b081-6a05e55da675" >
						<ee:message >
							<ee:set-payload ><![CDATA[%dw 2.0
output application/json
---
{
  traxion_response: {
    completed_succesfully: "false",
    error: {
      error_type: "CCA:" ++ (vars.main_error.identifier default "ERROR"),
      payload: {
      	event: correlationId
      },
      user_error_description: "error en comunicacion con CCA",
      system_error_description: vars.causeby
    }
  }
}]]></ee:set-payload>
						</ee:message>
						<ee:variables >
							<ee:set-variable variableName="slackChannel" ><![CDATA[%dw 2.0
output application/json
---
p('slack.connection.channel.connectivity.error')]]></ee:set-variable>
							<ee:set-variable variableName="messageProvider" ><![CDATA[%dw 2.0
output application/json
---
"ERROR PROVENIENTE DE WMS BLUEYONDER"]]></ee:set-variable>
						</ee:variables>
					</ee:transform>
					<flow-ref doc:name="orchestrator-send-message-to-slack-main" doc:id="7fd98624-00be-4a00-982f-0e0042989147" name="orchestrator-send-message-to-slack-main" />
					<raise-error doc:name="Raise error" doc:id="1f92693f-22c8-405d-b2c4-2359d95e8d8a" type="CCA:ERROR" description="#[vars.causeby]" />
				</on-error-continue>
				</error-handler>
			</try>
		<ee:transform doc:name="Catch http status 206" doc:id="d5ad94ff-605d-4938-bdde-27150c84e5c1">
				<ee:message>
				</ee:message>
				<ee:variables>
				<ee:set-variable variableName="codeErrorCca" ><![CDATA[attributes.statusCode]]></ee:set-variable>
                </ee:variables>
            </ee:transform>
        <validation:is-false doc:name="Is false" doc:id="915a2e37-0562-42eb-bc2f-8c039a2268ab" expression='#[(!isEmpty(payload[0].erroresM)) or (!isEmpty(payload[0].erroresD))]' message='#[write(flatten(payload.erroresM ++ payload.erroresD),"application/json")]'>
			<error-mapping sourceType="VALIDATION:INVALID_BOOLEAN" targetType="CCA:BAD_REQUEST" />
		</validation:is-false>
		<logger level="INFO" doc:name="Logger" doc:id="b0a9fc5c-af14-4524-b80e-c1f605a5af59" message="#[payload]"/>		

	</sub-flow>
</mule>
